### DISCUSSION TITLE
클로저가 활용되는 예시를 코드로 작성해주세요

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#80] [클로저]
> 클로저가 활용되는 예시를 코드로 작성해주세요
>  - 해당 코드에 대해 설명해주세요.

**Comment by publdaze** [comment]
> - 클로저가 활용되는 예시를 코드로 작성해주세요
```js
const counter = (function() {
  let cnt = 0;
  return {
    increase: () => cnt++,
    decrease: () => cnt--,
  }
}());

counter.increase();
counter.decrese();
```
>  - 해당 코드에 대해 설명해주세요.
- 즉시 실행 함수 코드가 실행되면서 cnt가 즉시 실행 함수의 렉시컬 환경에 저장
- 반환되는 객체의 메소드는 cnt 식별자를 참조
- 즉시 실행 함수가 종료되고 생명주기가 끝났는데,
- 반환된 메소드들이 cnt 식별자를 계속 참조하므로 렉시컬 환경은 남아있게 되고 참조가 가능
- 따라서 increase 메소드가 동작하면 cnt가 1로 변하고,
- 동일한 렉시컬 환경을 참조하는 decrease 메소드가 동작하면 1이었던 cnt가 0으로 다시 변함

**Comment by pipisebastian** [comment]
```js
const counter = (function(){
    let num = 1;
    return {
        add(){
            ++num;
        },
        minus(){
            --num;
        },
        print(){
            return num;
        }
    }
}());

counter.add();
console.log(counter.print());
```

- 즉시실행함수로 감쌈. 그래서 즉시실행함수는 실행컨텍스트에서 바로 빠져나오지만, return하는 객체의 함수들에서, 즉시실행함수의 식별자를 참조하고 있음. 그렇기에 지역변수 num은 렉시컬환경에 계속 살아있게됨.
